packages/engine/lib/error/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+///
+/// ```
+/// # struct MyErrorKind;
+/// type Result<T, E = error::Report<MyErrorKind>> = error::Result<T, E>;
+/// ```
+pub type Result<T, E = Report> = core::result::Result<T, E>;
+
+/// Provides error information for a [`Frame`].
+///
+/// In comparison to [`Context`], `ErrorKind` is not implemented for [`Display`] automatically. It
+/// can be used to force the user to provide more context information.
+///
+/// See the [`tags`] submodule for built-in [`TypeTag`]s used by this API.
+///
+/// [`TypeTag`]: provider::TypeTag
+/// [`Display`]: core::fmt::Display
+///
+/// ```
+/// use core::fmt;
+///
+/// use error::ErrorKind;
+/// use provider::Requisition;
+///
+/// # struct User;
+/// # struct Resource;
+/// # impl fmt::Display for User { fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { Ok(()) }}
+/// # impl fmt::Display for Resource { fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { Ok(()) }}
+/// struct PermissionDenied {
+///     user: User,
+///     resource: Resource,
+/// }
+///
+/// impl fmt::Display for PermissionDenied {
+///     fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+///         write!(fmt, "Permission denied accessing {} as {}", self.resource, self.user)
+///     }
+/// }
+///
+/// impl ErrorKind for PermissionDenied {
+///     fn provide<'p>(&'p self, req: &mut Requisition<'p, '_>) {
+///         req.provide_ref(&self.user).provide_ref(&self.resource);
+///     }
+/// }
+/// ```
+pub trait ErrorKind: fmt::Display + Send + Sync + 'static {
+    /// Provides error information for a [`Frame`] similar to the [`provider`] API.
+    fn provide<'p>(&'p self, req: &mut Requisition<'p, '_>);
+}
+
+impl<P> ErrorKind for P
+where
+    P: Provider + fmt::Display + Send + Sync + 'static,
+{
+    fn provide<'p>(&'p self, req: &mut Requisition<'p, '_>) {
+        Provider::provide(self, req);
+    }
+}
+
+/// Provides error information for a [`Frame`].
+///
+/// In comparison to [`ErrorKind`], `Context` is implemented for [`Display`] automatically, so even
+/// simple data like [`&str`][str] can be used to provide a better error context.
+///
+/// See the [`tags`] submodule for built-in [`TypeTag`]s used by this API.
+///
+/// [`TypeTag`]: provider::TypeTag
+/// [`Display`]: core::fmt::Display
+///
+/// # Example
+///
+/// You can create new context objects by implementing this trait, however the `min_specialization`
+/// feature has to be enabled:
+///
+/// ```
+/// #![feature(min_specialization)]
+///
+/// use core::fmt;
+///
+/// use error::Context;
+/// use provider::Requisition;
+///
+/// struct MyContext;
+///
+/// impl fmt::Display for MyContext {
+///     fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+///         fmt.write_str("Context information")
+///     }
+/// }
+///
+/// impl Context for MyContext {
+///     fn provide<'p>(&'p self, req: &mut Requisition<'p, '_>) {}
+/// }
+/// ```
+pub trait Context: fmt::Display + Send + Sync + 'static {
```

Comment:
Do we want to require `Context: Debug`?

Reply:
See above for my response for `ErrorKind` unless there's a some large distinction between these two types that I'm missing